sys/mips/mips/pmap.c

```diff
@@ -2618,10 +2618,9 @@ pmap_copy_page_internal(vm_page_t src, vm_page_t dst, int flags)
 		    MIPS_PHYS_TO_DIRECT(phys_dst), PAGE_SIZE);
 		va_src = MIPS_PHYS_TO_DIRECT(phys_src);
 		va_dst = MIPS_PHYS_TO_DIRECT(phys_dst);
-#ifdef CPU_CHERI
-		if (flags & PMAP_COPY_TAGS)
-			cheri_bcopy((caddr_t)va_src, (caddr_t)va_dst,
-			    PAGE_SIZE);
+#if __has_feature(capabilities)
+		if ((flags & PMAP_COPY_TAGS) == 0)
+			bcopynocap((caddr_t)va_src, (caddr_t)va_dst, PAGE_SIZE);
```

Comment:
If this is the only use of bcopynocap(), then maybe just add a new function in pmap.c instead?  That is what I did for risc-v.  Since it knows it's copying a page you can assume alignment and a constant number of loop iterations.  (See `pagecopy_cleartags`).  This avoids needing the new symbol and changes to bcopy.c, and it's actually not that unusual to have MD versions for page zeroing and page copying (e.g. using vector instructions on x86).

Reply:
It is the only use. Currently the use of nocap variants is very limited, it seems that they are only used by copyin/copyout.

Reply:
Yes.  For RISC-V copyin/out and copyin/outcap just use open-coded loops, though they use an assembly macro to reduce duplication.  But in particular copyin/out don't call the nocap versions directly (and this weirdness also means that for MIPS the nocap versions have to be in assembly instead of in C according to comments in the MIPS assembly)

Reply:
Anyway, this specific use of bcopynocap can assume page alignment but the others below can not. This is annoying as I would and up producing another copy of the libkern code here. As an aside `pmap_copy_pages_tags` is never used, which is why the RISC-V version gets away without handling it. Perhaps we should remove it altogether.

Reply:
Oh, I implemented it for RISC-V:

https://github.com/CTSRD-CHERI/cheribsd/blob/master/sys/riscv/riscv/pmap.c#L335

The kernel didn't link otherwise.  My suggestion is to just grab that function and use it in the MIPS pmap as well and then drop bcopynocap entirely.

Reply:
If you made `src` a capability you could clear the `LOAD_CAPABILITY` permission bit and get tag stripping for free :) ...

Reply:
Oh, I see, pmap_copy_pages_tags() and pmap_copy_pages().   Hmmm, RISC-V is currently preserving tags then in pmap_copy_pages() when it perhaps should not be?  If we aren't using pmap_copy_pages_tags() then removing it may indeed be good, but then pmap_copy_pages() in RISC-V's pmap also needs to use bcopynocap().

Reply:
@bsdjhb Yes I think that the RISC-V version should also use the tag stripping version instead, I got it backwards in the previous comment.

Reply:
@bsdjhb If it is to be used by RISC-V as well I can keep the current bcopynocap implementation and add the RISC-V changes here.

Reply:
Yes, sorry, that was the intention of my last comment.